File: baekjoon/16236/16236.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Main {
    private static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    private static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    private static int N;
    private static int[][] matrix;
    private static Location[] adjoinCriteria = {
            new Location(-1, 0),
            new Location(0, -1),
            new Location(1, 0),
            new Location(0, 1)
    };
    private static int result;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(reader.readLine());
        matrix = new int[N][N];

        for (int i = 0; i < N; i++) {
            matrix[i] = toIntArray(reader.readLine().split(" "));
        }

        solve();

        writer.write(String.format("%d\n", result));
        writer.flush();

        writer.close();
    }

    private static int[] toIntArray(String[] strings) {
        int[] arr = new int[strings.length];

        for (int i = 0; i < strings.length; i++) {
            arr[i] = Integer.parseInt(strings[i]);
        }

        return arr;
    }

    /*
       요구사항

        [이동]
        - 자신의 크기보다 큰 물고기가 있는 칸은 이동 불가, 그 이외에는 이동 가능

        [먹기]
        - 자신의 크기보다 작은 물고기만 먹을 수 있음
        - 자신의 크기와 같은 수의 물고기를 먹을때마다 크기가 1 증가

        [먹을 물고기 선정 방법]
        - 먹을 수 있는 물고기가 공간에 없다면 엄마 상어에게 도움을 요청
        - 먹을 수 있는 물고기가 1마리라면 그 물고기를 먹으러 감
        - 먹을 수 있는 물고기가 2마리 이상이라면 가장 가까운 물고기를 먹으러 감
           - 가까움의 기준
             - 지나가야 하는 칸의 개수
             - 가장 가까운 물고기가 여러개라면
               - 가장 위에 있는 물고기 우선
               - 그 다음 가장 왼쪽에 있는 물고기 우선
     */

    private static void solve() {
        Location babySharkLocation = null;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N ; j++) {
                if (matrix[i][j] == 9) {
                    babySharkLocation = new Location(i, j);
                    break;
                }
            }
        }

        Shark babyShark = new Shark(babySharkLocation, 2);

        boolean eaten = true;
        int eatenCount = 0;
        while (eaten) {
            eaten = bfs(babyShark, babyShark.size);

            if (eaten) {
                eatenCount++;
            }
            if (eatenCount >= babyShark.size) {
                babyShark.size++;
                eatenCount = 0;
            }
        }
    }

    private static boolean bfs(Shark babyShark, int babySharkSize) {
        boolean[][] visited = new boolean[N][N];
        visited[babyShark.location.row][babyShark.location.col] = true;
        Queue<Information> queue = new LinkedList<>();
        queue.add(new Information(new Location(babyShark.location.row, babyShark.location.col), 0));

        List<Information> eatableInformations = new ArrayList<>();

        while (!queue.isEmpty()) {
            Information currentInformation = queue.poll();

            for (Location adjoinCriterion : adjoinCriteria) {
                Location adjoinLocation = new Location(currentInformation.location.row + adjoinCriterion.row, currentInformation.location.col + adjoinCriterion.col);

                // 이동 가능하고 방문하지 않는 경우만 고려
                if (!canMoveTo(adjoinLocation, babySharkSize) || visited[adjoinLocation.row][adjoinLocation.col]) {
                    continue;
                }

                // 먹을 수 있다면
                if (canEat(adjoinLocation, babySharkSize)) {
                    visited[adjoinLocation.row][adjoinLocation.col] = true;
                    eatableInformations.add(new Information(new Location(adjoinLocation.row, adjoinLocation.col), currentInformation.distance + 1));
                } else {
                    visited[adjoinLocation.row][adjoinLocation.col] = true;
                    queue.add(new Information(new Location(adjoinLocation.row, adjoinLocation.col), currentInformation.distance + 1));
                }
            }
        }

        if (eatableInformations.isEmpty()) {
            return false;
        }

        eatableInformations.sort(Information::compareTo);
        Information nextBabySharkInformation = eatableInformations.get(0);
        matrix[babyShark.location.row][babyShark.location.col] = 0;
        matrix[nextBabySharkInformation.location.row][nextBabySharkInformation.location.col] = 9;
        babyShark.location.row = nextBabySharkInformation.location.row;
        babyShark.location.col = nextBabySharkInformation.location.col;
        result += nextBabySharkInformation.distance;

        return true;
    }

    private static boolean canMoveTo(Location location, int babySharkSize) {
        if (location.row < 0 || location.row >= N || location.col < 0 || location.col >= N) {
            return false;
        }

        return matrix[location.row][location.col] <= babySharkSize;
    }

    private static boolean canEat(Location location, int babySharkSize) {
        if (matrix[location.row][location.col] == 0) {
            return false;
        }

        return matrix[location.row][location.col] < babySharkSize;
    }

    private static class Location {
        int row;
        int col;

        public Location(int row, int col) {
            this.row = row;
            this.col = col;
        }
    }

    private static class Information {
        Location location;
        int distance;

        public Information(Location location, int distance) {
            this.location = location;
            this.distance = distance;
        }

        public int compareTo(Information information) {
            if (this.distance != information.distance) {
                return this.distance - information.distance;
            } else if (this.location.row != information.location.row) {
                return this.location.row - information.location.row;
            } else {
                return this.location.col - information.location.col;
            }
        }
    }

    private static class Shark {
        Location location;
        int size;

        public Shark(Location location, int size) {
            this.location = location;
            this.size = size;
        }
    }
}

File: baekjoon/16236/16236.py
import sys
from collections import deque
from functools import cmp_to_key

class Location:
    def __init__(self, row, col):
        self.row = row
        self.col = col

class BFSInformation:
    def __init__(self, location, distance):
        self.location = location
        self.distance = distance

class Shark:
    def __init__(self, location, size):
        self.location = location
        self.size = size

matrix = []
N = 0
adjacent_directions = [
    Location(-1, 0),
    Location(0, -1),
    Location(1, 0),
    Location(0, 1)
]
result = 0

def solve():
    baby_shark_location = None

    for i in range(N):
        for j in range(N):
            if matrix[i][j] == 9:
                baby_shark_location = Location(i, j)
                break

    baby_shark = Shark(baby_shark_location, 2)

    eaten = True
    eaten_count = 0
    while eaten:
        eaten = bfs(baby_shark)
        if eaten:
            eaten_count += 1
        if eaten_count >= baby_shark.size:
            baby_shark.size += 1
            eaten_count = 0

def bfs(baby_shark):
    global result

    # 방문 배열 setting
    visited = [[False for _ in range(N)] for _ in range(N)]
    visited[baby_shark.location.row][baby_shark.location.col] = True

    # queue 세팅
    queue = deque()
    queue.append(BFSInformation(Location(baby_shark.location.row, baby_shark.location.col), 0))

    # 먹을 수 있는 물고기들에 대한 bfs 정보 list
    eatable_fish_informations = []

    while bool(queue):
        current_bfs_information = queue.popleft()

        for adjoin_direction in adjacent_directions:
            adjoin_location = Location(current_bfs_information.location.row + adjoin_direction.row, current_bfs_information.location.col + adjoin_direction.col)

            # 이동이 불가능하거나 이미 방문한 위치라면 pass
            if not can_move_to(adjoin_location, baby_shark.size) or visited[adjoin_location.row][adjoin_location.col]:
                continue

            # 지금부턴 이동이 가능한 경우임.

            visited[adjoin_location.row][adjoin_location.col] = True

            # 먹을 수 있는 물고기가 위치한 곳이라면 list에 추가
            if can_eat(adjoin_location, baby_shark.size):
                eatable_fish_informations.append(BFSInformation(Location(adjoin_location.row, adjoin_location.col), current_bfs_information.distance + 1))
            else:
                queue.append(BFSInformation(Location(adjoin_location.row, adjoin_location.col), current_bfs_information.distance + 1))

    # 먹을 수 있는 물고기가 없으면 없다고 알림
    if len(eatable_fish_informations) == 0:
        return False

    # 아래에서부턴 먹을 수 있는 물고기가 존재하는 경우임
    eatable_fish_informations.sort(key=cmp_to_key(compare))
    next_baby_shark_information = eatable_fish_informations[0]

    # 최신화
    matrix[baby_shark.location.row][baby_shark.location.col] = 0
    matrix[next_baby_shark_information.location.row][next_baby_shark_information.location.col] = 0
    baby_shark.location.row = next_baby_shark_information.location.row
    baby_shark.location.col = next_baby_shark_information.location.col

    # 시간(거리) 누적
    result += next_baby_shark_information.distance

    # 먹을 수 있는 물고기가 있었다고 알림
    return True

def can_move_to(location, baby_shark_size):
    if location.row < 0 or location.row >= N or location.col < 0 or location.col >= N:
        return False

    return matrix[location.row][location.col] <= baby_shark_size

def can_eat(location, baby_shark_size):
    if matrix[location.row][location.col] == 0:
        return False

    return matrix[location.row][location.col] < baby_shark_size

def compare(bfs_information1, bfs_information2):
    if bfs_information1.distance != bfs_information2.distance:
        return bfs_information1.distance - bfs_information2.distance
    elif bfs_information1.location.row != bfs_information2.location.row:
        return bfs_information1.location.row - bfs_information2.location.row
    else:
        return bfs_information1.location.col - bfs_information2.location.col

if __name__ == '__main__':
    N = int(sys.stdin.readline())
    for _ in range(N):
        matrix.append(list(map(int, sys.stdin.readline().split())))

    solve()

    print(result)
